### ISSUE TITLE
vgg weights in torch and keras

### ISSUE THREAD

**Issue by betterze** [moono/lpips-tf2.x#2]
Dear Moono,

Thank you @moono for sharing with us this implementation.

If i understand correctly, the weights in keras vgg are taken from first vgg model (trained in caffe), and the weights in torch vgg is trained by torch team. Since the  weights and image preprocessing methods are different, I am not sure we can use the 'lin' weights that trained on torch vgg. 

Is my understanding correct?

Thank you for your help.

Best Wishes,

Alex

**Comment by moono** [comment]
Hello,

This repo doesn't contains how to train 'lin' weights, just copied whole weights from official pytorch repo.
If you see `convert_to_tensorflow.py` you will understand how I converted the weights.
My only assumption is that network architecture vgg from keras is same as pytorch one.

To summarize how I converted to tf follows,
* load official lpips model (pytorch)
* hold on to pytorch version weights (vgg & lin) as numpy array
* create tensorflow lpips model
* copy the official pytorch numpy weights to tensorflow lpips model while matching shapes
* save the tensorflow lpips model as *.ckpt format, so next time we don't have to use pytorch anymore.
* check if saved(exported) tensorflow lpips model outputs as same as pytorch one(`test.py`) --> passed

If you are worried about vgg weights loading(downloading) at the beginning when creating tensorflow lpips model, it's okay to ignore. Because those weight will be overridden when loading with *.ckpt file.

**Comment by betterze** [comment]
Dear Moono,

Thank you @moono for your detailed reply. I understand it now. 

I am trying to use this LPIPS loss to train an auto-encoder. In the first few iteratiation, it runs well, then suddenly the loss becomes inf and encoder outputs nan. I verify that in the begining, encoder outputs valid value and the loss is around 0.8. This (nan loss and encoder outputs nan) never happens when I use regular vgg loss.

I check the code of LPIPS, I could not find things that can cause gradient explosion. Would you mind telling me why this happen? 

Thank you very much for your help.

Best Wishes,

Alex

**Comment by moono** [comment]
Well, sorry but I'm not sure why gradient explosion happens...
I used this repo's lpips model to optimize (encode) stylegan2 model and it worked fine.
link: https://github.com/moono/stylegan2-encoder

**Comment by betterze** [comment]
@moono Thx, I will take a lot of this repo.
